### PR TITLE
Initial config for TypeScript refactor

### PR DESCRIPTION
**Pull request by hleote**
Addresses #516

## Purpose
- Add initial config for refactoring components to TypeScript

## Approach and changes
- Added initial config files and partially refactored Button component to TypeScript

## Definition of done

* [ ] Development completed
* [x] Reviewers assigned
* [ ] Unit and integration tests
* [ ] Meets minimum browser support
* [ ] Meets accessibility requirements

### CODE REVIEW

File: src/tsconfig.json [hunk 1]
```diff
@@ -0,0 +1,65 @@
+{
+    "compilerOptions": {
```
Comment (robinmetral) [inline src/tsconfig.json]: Note: when we ship this, we can remove `jsconfig.json` (see #520) and merge its config into this file.

File: src/components/Button/Button.ts [hunk 2]
```diff
@@ -56,60 +55,36 @@ const REGULAR_BUTTON_ONLY_PROPS = [
...
 ];
 
+enum SIZE_PROP_TYPE {
+  KILO,
+  MEGA,
+  GIGA
+}
+
+interface Props {
+  size: SIZE_PROP_TYPE // Size of the button. Use the Button's KILO, MEGA, or GIGA properties.
```
Comment (connor-baer) [inline src/components/Button/Button.ts:65]: The comments need to remain this format so they can be picked up by [react-docgen](https://github.com/reactjs/react-docgen).

```suggestion
  /**
   * Size of the button. Use the Button's KILO, MEGA, or GIGA properties.
   */
  size: SIZE_PROP_TYPE 
```

File: src/components/Button/Button.ts [hunk 3]
```diff
@@ -56,60 +55,36 @@ const REGULAR_BUTTON_ONLY_PROPS = [
...
+  href: string // URL the Button should lead to. Causes the Button to render an <a> tag.
+  flat: boolean // Button has a 'flat' variation, triggered with this prop.
+  plain: boolean // Makes the button look and behave like a text link.
+  target: string // Link target. Should only be passed, if href is passed, too.
+  stretch: boolean // Trigger stretch (full width) styles on the component.
+  primary: boolean // Renders a primary button using the brand color.
+  disabled: boolean // Should the Button be disabled?
+  secondary: boolean // Renders a secondary button. Secondary buttons look the same for primary (default) and flat buttons.
+  defaultProps: {}
```
Comment (connor-baer) [inline src/components/Button/Button.ts:74]: This is not a prop type 😉

File: src/components/Button/Button.ts [hunk 4]
```diff
@@ -118,11 +93,15 @@ Button.defaultProps = {
...
   secondary: false,
-  size: Button.MEGA,
+  size: MEGA,
   stretch: false,
   target: null
 };
 
+Button.KILO = KILO;
+Button.MEGA = MEGA;
+Button.GIGA = GIGA;
```
Comment (connor-baer) [inline src/components/Button/Button.ts:103]: Do you know how to type static properties on a function? I couldn't find anything when I googled it 🤔

File: src/tsconfig.json [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+      "noEmit": true,
```
Comment (connor-baer) [inline src/tsconfig.json:3]: We want to use TypeScript as our compiler, not just for type checking.

File: src/tsconfig.json [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+      /* Basic Options */
+      // "incremental": true,                   /* Enable incremental compilation */
+      "target": "es5",                          /* Specify ECMAScript target version: 'ES3' (default), 'ES5', 'ES2015', 'ES2016', 'ES2017', 'ES2018', 'ES2019' or 'ESNEXT'. */
+      "module": "commonjs",                     /* Specify module code generation: 'none', 'commonjs', 'amd', 'system', 'umd', 'es2015', or 'ESNext'. */
```
Comment (connor-baer) [inline src/tsconfig.json:7]: This needs to be `es2015` or higher, otherwise, tree-shaking won't work.

File: src/tsconfig.json [hunk 7]
```diff
@@ -0,0 +1,65 @@
+{
+    "compilerOptions": {
+      "noEmit": true,
+      /* Basic Options */
+      // "incremental": true,                   /* Enable incremental compilation */
+      "target": "es5",                          /* Specify ECMAScript target version: 'ES3' (default), 'ES5', 'ES2015', 'ES2016', 'ES2017', 'ES2018', 'ES2019' or 'ESNEXT'. */
```
Comment (connor-baer) [inline src/tsconfig.json:6]: We want to publish modern code so the user can decide if they need to transpile the code. Next.js and CRA now transpile node modules, and Next.js even provides modern bundles for newer browsers.

**Comment by connor-baer** [comment]
Closing in favor of #563.

**Pull request closed without merging**